drivers/usb/udc/udc_mcux.c

```diff
@@ -0,0 +1,903 @@
+	.disable = usb_mcux_disable,
+	.init = usb_mcux_init,
+	.shutdown = usb_mcux_shutdown,
+	.lock = usb_mcux_lock,
+	.unlock = usb_mcux_unlock,
+};
+
+#ifdef CONFIG_USB_UDC_NXP_EHCI
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT nxp_ehci
+
+
+#ifdef CONFIG_USB_UDC_NXP_PHY
+#define UDC_MCUX_PHY_CONFIGURE(n)							\
+	.need_to_init_phy = DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle) ? 1u : 0u,	\
+	.D_CAL = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-d-cal, 0),		\
+	.TXCAL45DP = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dp-ohms, 0),	\
+	.TXCAL45DM = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dm-ohms, 0),
+#else
+#define UDC_MCUX_PHY_CONFIGURE(n)
+#endif
+
+#define USB_MCUX_EHCI_DEVICE_DEFINE(n)							\
+	static void udc_irq_enable_func##n(const struct device *dev)			\
+	{										\
+		IRQ_CONNECT(DT_INST_IRQN(n),						\
+			    DT_INST_IRQ(n, priority),					\
+			    udc_mcux_ehci_isr,						\
+			    DEVICE_DT_INST_GET(n), 0);					\
+											\
+		irq_enable(DT_INST_IRQN(n));						\
+	}										\
+											\
+	static void udc_irq_disable_func##n(const struct device *dev)			\
+	{										\
+		irq_disable(DT_INST_IRQN(n));						\
+	}										\
+											\
+	static struct udc_ep_config							\
+		ep_cfg_out##n[DT_INST_PROP(n, num_bidir_endpoints)];			\
+	static struct udc_ep_config							\
+		ep_cfg_in##n[DT_INST_PROP(n, num_bidir_endpoints)];			\
+											\
+	PINCTRL_DT_INST_DEFINE(n);							\
+											\
+	static struct udc_mcux_config priv_config_##n = {				\
+		UDC_MCUX_PHY_CONFIGURE(n)						\
+		.base = (uint32_t)DT_INST_REG_ADDR(n),					\
+		.irq_enable_func = udc_irq_enable_func##n,				\
+		.irq_disable_func = udc_irq_disable_func##n,				\
+		.num_of_eps = DT_INST_PROP(n, num_bidir_endpoints),			\
+		.ep_cfg_in = ep_cfg_in##n,						\
+		.ep_cfg_out = ep_cfg_out##n,						\
+		.mcux_if = &mcux_ehci_usb_iface,					\
+		.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),				\
+	};										\
+											\
+	static struct udc_mcux_data priv_data_##n = {					\
+	};										\
+											\
+	static struct udc_data udc_data_##n = {						\
+		.mutex = Z_MUTEX_INITIALIZER(udc_data_##n.mutex),			\
+		.priv = &priv_data_##n,							\
+	};										\
+											\
+	DEVICE_DT_INST_DEFINE(n, usb_mcux_driver_preinit, NULL,				\
+			      &udc_data_##n, &priv_config_##n,				\
+			      POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		\
+			      &usb_mcux_api);
+
+DT_INST_FOREACH_STATUS_OKAY(USB_MCUX_EHCI_DEVICE_DEFINE)
+#endif
+
+#ifdef CONFIG_USB_UDC_NXP_IP3511
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT nxp_lpcip3511
+
+
+#ifdef CONFIG_USB_UDC_NXP_PHY
+#define UDC_MCUX_PHY_CONFIGURE(n)							\
+	.need_to_init_phy = DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle) ? 1u : 0u,	\
+	.D_CAL = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-d-cal, 0),		\
+	.TXCAL45DP = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dp-ohms, 0),	\
+	.TXCAL45DM = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx-cal-45-dm-ohms, 0),
```

Comment:
This can be improved, I will comment on it once you have addressed other comments.